tools/expand_stamp_vars_test.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----
+
+# A test for the filter that expands Bazel workspace stamp variables.
+
+from tools import expand_stamp_vars
+
+import io
+import sys
+
+if __name__ == "__main__":
```

Comment:
Is it possible to write this test in the form of a python unittest?

Reply:
Sure thing; although as I do, please help me articulate a general approach.

The only precedent I saw for Python-language unit tests in our tree is the use of `tensorflow.python.test`, from the `tensorflow` package. For example:

https://github.com/tensorflow/tflite-micro/blob/46ccaf8fb3687d0f6b579f726400f81ab5fa7a81/python/tflite_micro/runtime_test.py#L27-L28
https://github.com/tensorflow/tflite-micro/blob/46ccaf8fb3687d0f6b579f726400f81ab5fa7a81/python/tflite_micro/runtime_test.py#L33-L34
https://github.com/tensorflow/tflite-micro/blob/46ccaf8fb3687d0f6b579f726400f81ab5fa7a81/python/tflite_micro/runtime_test.py#L268-L269

The layers here are something like:

1. `tensorflow.python.test` [wraps](https://github.com/tensorflow/tensorflow/blob/f52dc5dafc43fabc594f1c1bf744c0ae585ee4bc/tensorflow/python/platform/test.py#L24)
2. `absl.testing.absltest` ([via](https://github.com/tensorflow/tensorflow/blob/f52dc5dafc43fabc594f1c1bf744c0ae585ee4bc/tensorflow/python/platform/googletest.py#L25) the name `testing.pybase.googletest`), which [wraps](https://github.com/abseil/abseil-py/blob/976413320682f8fd1c05c36f24a9475050702699/absl/testing/absltest.py#L564)
3. the standard library's `unittest`. All run as a
4. Bazel `py_test()`, which like all Bazel tests, simply expects an executable that returns success or failure.

Because what I'm testing here is just a scrap of generic build tooling, it didn't seem right to intertwine a heavy dependency like `tensorflow`. It seemed so lite, in fact—it's trivially short, I'm not testing multiple cases, I'm not aware of any time we run things like this from outside Bazel (where it might benefit from the discoverability of `unittest`), etc.—I just jumped in at no. 4 there, the lowest layer, and wrote a straightforward script that returns success or failure.

I imagine the rationale for using `unittest` being: it's standard so it will cue the reader that this is a test—even though I can't think of any features of `unittest` from which it will benefit. I myself have some room for low-level, Bazel-ly test executables that simply pass or fail, like I see in other Bazel-adjacent projects like [bazelbuild/rules_python](https://github.com/bazelbuild/rules_python/tree/main).

Would there be reasons for going even higher up the stack? If one accepts the argument for `unittest`, I could imagine an argument for `absltest` as The Google Way to do `unittest`. I strongly dislike the idea of bringing in `tensorflow` for this. Again, remember this is just a small, stand-alone test of some very Bazel-centric glue code.

Maybe there are some motivations or requirements I don't see?

Reply:
Refactored to use `unittest` in dd643ed5.

Reply:
The motivation is commonality. Ideally, tests are written in one framework so that developers don't have to be familiar with more than one. Perhaps I should have suggested `tensorflow.python.test` to get closer to the norm (I'm still new around here). While I agree this is a pretty small test and that adding dependencies should be done thoughtfully, there is still value in commonality.